### PR TITLE
New AWS SES module to enable DKIM on an identity…

### PR DESCRIPTION
**Pull request by stefanhorning**
… and/or start DKIM verification process.

##### SUMMARY
New module aws_ses_identity_dkim for handling DKIM settings on an new or existing SES identity.

See https://docs.aws.amazon.com/ses/latest/DeveloperGuide/easy-dkim-setup-domain.html

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
aws_ses_identity_dkim

##### ADDITIONAL INFORMATION

For usage see module docs. This module complements other, already existing SES modules, like `aws_ses_identity` and `aws_ses_policy`.

### CODE REVIEW

**Comment by stefanhorning** [comment]
Any reason not to merge this?

**Comment by orthanc** [comment]
Thanks for the PR @stefanhorning.

Can you add an integration test for this module please. There's a guide here on how to do that https://docs.ansible.com/ansible/devel/dev_guide/platforms/aws_guidelines.html#integration-tests-for-aws-modules (the guides in the progress of being moved from a github readme to ansible docs proper which is why its only on the dev branch docsite). If you need some examples to start with there are existing tests for the other aws_ses_* modules under `test/integration/targets`.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: To make the module support check_mode it would be good to wrap this with

    if not module.check_mode:

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: To make the module support check_mode it would be good to wrap this with

    if not module.check_mode:

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: Stylisic, but I'd prefer to return from the previous line rather than assigning to a variable that's immediately returned

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: To make the module support check_mode it would be good to wrap this with

    if not module.check_mode:

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:213]: Can we adopt the same change checking strategy used in the other SES modules (e.g. https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py#L113)

Basically:

* Call get_identity_dkim_settings BEFORE this point
* Check if the desired state is different from the current state, if so, set changed=True and call the relevant methods to make it so
* If chnaged is true, call get_identity_dkim_settings AGAIN below where you have it to get the updated settings to return

Doing it this way works well because:
* We don't call APIs to change things if there's no need
* The module is idempotent and correctly sets the changed state
* check_mode can be easily implemented by just skipping the mutation operations (see other review comments)

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: changed needs to be set to True if we disable DKIM since that's still a change, see comment above on how best to do that.

**Review by orthanc — CHANGES_REQUESTED** [review]
A few detail comments from reviewing the code, haven't actually run this module.

Key things from code review though:
* The module is not correctly handling the changed flag, theres an existing pattern in the other SES modles that could be adopted here I think
* The module declaes that check mode is supported but doesn't implement it. Detail notes in comments on how to make it supported.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: Stylistic, but since we're not doing anything sith the response here, prehaps just return directly on line 133 rather than assigning to a varible and returning after the catch

**Comment by stefanhorning** [comment]
Thanks @orthanc for the code review, I have addressed all issues except the change detection now.

I have also addded an integration test that still needs a CI update to allow the IAM action `ses:VerifyDomainDkim`. Once that's fixed it will probably still be red until I implement a proper change detection as mentioned above. 
Will probably get arround to do it in the coming days, in the meantime it would be nice if s.b. could already look into the CI update.

File: hacking/aws_config/testing_policies/ses-policy.json
Comment (orthanc) [inline hacking/aws_config/testing_policies/ses-policy.json]: For some reasons the rest of the SES permissions are in the [compute-policy.json](https://github.com/ansible/ansible/blob/devel/hacking/aws_config/testing_policies/compute-policy.json#L227)

I like the idea of pulling the ses permissions out into it's own policy file since it's a bit of a different area if we're going to do this it would be good to move the rest of the SES permissions into ses-policy.json as well.

I've asked on the #ansible-aws irc channel if there's a reason to keep these in compute instead.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: It would be really nice if we could limit this to a specific error code or something. I can't see anything in the AWS docs about a specific error for this case. I'll have a poke and see if I can identiify a way to do this.

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (orthanc) [inline test/integration/targets/aws_ses_identity_dkim/tasks/main.yml:129]: It would be good to add a re-run version of the 2 diable checks as well to confirm idempotence on disable as well as enable.

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (orthanc) [inline test/integration/targets/aws_ses_identity_dkim/tasks/main.yml:191]: It would be good to add a couple of tests for check_mode as well to make sure it doesn't actually get broken.

There's an example of how to do this here https://github.com/ansible/ansible/blob/devel/test/integration/targets/aws_ses_identity/tasks/main.yaml#L151

Essentially, you can specify

    check_mode: True

On the task to force a task into check mode, then of course you're asserting that it didn't change anything but reports that it did.

**Review by orthanc — CHANGES_REQUESTED** [review]
Thanks @stefanhorning 

Changes & tests look good. I've added a note about a couple of additional cases that it woulld be good to test for.

I think it would be a good idea to move all the SES permissions into the new SES test policy you've created, but I'll poke that one into the IRC channel and see if there's a reason for the rest of the permissions to be where they are.

We'll have to wait for one of the core team members to update the CI permissions, but I've found they're pretty good about picking those up once the PR is otherwise ready to go.

File: hacking/aws_config/testing_policies/ses-policy.json
Comment (willthames) [reply hacking/aws_config/testing_policies/ses-policy.json]: We can't have more than 10 policies (that's the most policies you can attach to a single group), so no new policies please (if you do want a new policy, have a look at combining existing policies first) - unless you want to add another group

Comment (orthanc) [reply hacking/aws_config/testing_policies/ses-policy.json]: Makes sense. Given that best to move these permissions into compute-policy.json with the rest of the SES permissions.

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (orthanc) [inline test/integration/targets/aws_ses_identity_dkim/tasks/main.yml]: The key here doesn't match the module.

The module currently returns all the atttributes wrapped in a dkim_attributes property so this would have to be:

    domain_resullt.dkim_attributes.dkim_enabled

But I'd suggest changing the module so it directly returns the attributes as I think the way it looks in this test is easier to read. 

That would require replacing line 168 in the module with

    module.exit_json(changed=changed, **camel_dict_to_snake_dict(attributes))

and updating the examples and return value documentation at the top of the module.

File: test/integration/targets/aws_ses_identity_dkim/defaults/main.yml
Comment (orthanc) [inline test/integration/targets/aws_ses_identity_dkim/defaults/main.yml]: These should be based on the resource_prefx to prevent parallel test runs interacting on the same identities. I'd suggest:

    dkim_identity_domain: "{{ resource_prefix }}-domain.example.com"
    dkim_identity_email: "address@{{ resource_prefix }}-email.example.com"

The use of resource_prefix will ensure that each test run is using fresh domains. This is often important as the CI build runs python2 / python3 runs in parallel so we don't want conflicting resources that would make the build unstable.

I suggest using a different sub-domain for the two identities because  there are interesting interactions between the domain and the email since we may just be enabling DKIM for a single email but DKIM is fundamentally domain based. So by using a different sub-domain for the two cases we remove any possibility that they interact.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: @stefanhorning I can't replicate the behaviour descibed in this comment. Runnig the tests without this catch seems to work fine. I wonder if AWS has updated the API to not error in this case. Can you chec if this catch and pass is still necessary?

Comment (stefanhorning) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: Couldn't reproduce it either anymore. So I removed this.

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_ses_identity_dkim/tasks/main.yml:129]: Done

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_ses_identity_dkim/tasks/main.yml:191]: Done

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_ses_identity_dkim/tasks/main.yml]: Decided to stay with the current logic and fix the tests instead. As I conciously decided to keep the AWS return values in it's own dict back then, just forgot about the structure when writing the test. ;)
It seems kind of a convention for (AWS) Ansible modules nowadays to seperate aws return values from the things returned by Ansible directly (kind of seperate namespacing) and I kind of like that approach as it allows to go one level deeper to filter out all Ansible stuff and still see all AWS stuff at a glance.

File: test/integration/targets/aws_ses_identity_dkim/defaults/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_ses_identity_dkim/defaults/main.yml]: Yep, makes sense. Done

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (stefanhorning) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:213]: Implemented now

**Comment by stefanhorning** [comment]
@orthanc I believe, I now addressed all your code review comments. Let me know if you are satisfied now.

File: test/integration/targets/aws_ses_identity_dkim/defaults/main.yml
Comment (orthanc) [inline test/integration/targets/aws_ses_identity_dkim/defaults/main.yml]: When running the tests locally ( `ansible-test integration aws_ses_identity_dkim` ) I'm still getting failures with this setup.

When you enable DKIM for a domain identity it enables it for any email identities within that domain. So the email identity test fails because:

1. DKIM is enabled by the domain identity test first
2. Email Identity tests don't actually change anything since DKIM is already enabled
3. Email identity tests fail because we were expecting a change and it didn't change.

Just changing this so the email identity and the domain identity are in different subdomains (still based off resource_prefix) solves the problem

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: I'm getting intermittennt failures running the tests. Occasionally the change that was just made is reflected in the returned properties.

I think what's happening is an eventual consistency issue between the varous SES APIs, as I've hit that with some of the other SES modues.

I think we need something similar to this retry loop here: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/aws_ses_identity.py#L234 setup to retry the get_identity_dkim_settings call (or possibly within the method) until the enabled status is what we expect and DKIM tokens exist (if enabling). This needs caution with check_mode though.

A slighly more fiddly approach that can work well is to actually update the previously loaded attributes with what you know is supposed to happen (example here https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py#L133 ). I prefer this approach when I can, but it would be fiddly to do here because you'd need to grab the DKIM tokens from the ses_verify_dkim_domain call.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py]: I've started seeing the Cliennt Error you were origionally catching here. It's reportig that we have't called verify_domain_dkim not that the verification is not yet complete (which is what I thought your comment was saying)

We definatly do call verify_domain_dkim and it only happens intermittently so I think there's another race condition in the AWS APIs like the one I mentioned on the get_attributes call.

This probablu also needs a retry loop on the ClientError to deal with the race condition.

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (orthanc) [inline test/integration/targets/aws_ses_identity_dkim/tasks/main.yml]: This test passes even if it's not run in check mode, so it's not ideal for testing check_mode

It would be better to test the check mode before the eable / after disable as that will fail if check mode actually makes the changes.

I've created this approach here if you want to just grab it: https://github.com/orthanc/ansible/commit/1cfa2166c868d718fd9a43693623f5499a538713

**Review by orthanc — CHANGES_REQUESTED** [review]

**Comment by orthanc** [comment]
@stefanhorning Thanks for the updates. Generally look good. I've picked up a couple of things from running the tests that I've suggested fixes for. Namely:
* using a different subdomain for email and domain identity tests
* A different way or doing check mode tests

Bigger thing though is the test's fail intermittently with repeated runs. I think this is due to delays in SES APIs reflecting chagnes due to internal eventual constency. I keep running into this when trying to write the SES modues. Pretty sure this was the cause of the mystery client error that I asked you to remove.

Best approach I've found so far is to wrap some elements in retry loops that look for a particular condition as well as the retry decorator that deals with throttteling.

I've made a start on this here: https://github.com/orthanc/ansible/commit/3753c346587cf87fc1540f7ee0090e2865e719da but the tests are still not fully reliable. So needs some more work. Hopefully you can pick it up otherwise I'll have a poke when I get time.

The key intermittent failure I'm still getting is:
1. Test enables or disables DKIM
2. This correctly returns the changed True and new attributes
3. The following re-run calls get_identity_dkim_settings and gets the old state so re-enables / disables and reports the wrong change status

File: test/integration/targets/aws_ses_identity_dkim/defaults/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_ses_identity_dkim/defaults/main.yml]: Yes, sorry oversight of mine. Fixed now

**Comment by stefanhorning** [comment]
If the issues only happen in the tests, because they run to quickly after each other, isn't it enough to just add some sleeps between the test tasks and be done with it? 
Or is there another issue I didn't get from your comment?

Fixed the domain settings now. Will look into the check-mode thing later.

**Comment by orthanc** [comment]
So with the retries / unstable tests....

The couple I patched and linked are probably worth including at least. While these are showing in tests the impact could actually happen and affect real module usage since the affect is to return the wrong result from the module which is not great. The retry loops I added should fix that reliably and so I'd say should just be included.

I've been doing a bit more investigating on the remaining instability and the result is confusing. I think we may need to do a test hack to work around it but I don't think a wait will do.

The remaining instability is because sometimes `get_identity_dkim_attributes` returns an out of date status even after we've looped to wait for the correct status. Even waiting for a minute doesn't seem to be enough to guarentee a consistent result.

My theory is that under the hood at AWS, the SES services are a cluster and it takes time for the new state to fully replicate, and sometimes we get a node that's got the new state, sometimes one with the old state. But that's just guesswork since AWS is a black box.

I know the core team really don't like to accept modules without tests or with flakey tests since it adds a maintence burden.

I think the best option here might be a "magic" envronment variable that allows the test to "force" an expected initial state so we get consistent behaviour. I've kinda got most of this as part of investigating the instablity so I'll tidy it up and push to my branch so you can take a look and see what you think.

**Comment by stefanhorning** [comment]
Alright, let me know when you are done, then I will pull in your changes.

If it's easier for you we can also open a PR based on your branch, as I think you are much more into the details of this module at the moment...

**Comment by orthanc** [comment]
@stefanhorning I've pushed the changes I think are needed here https://github.com/orthanc/ansible/tree/new_module_aws_ses_dkim there are 3 changes:

1. Retry loops on get / set dkim attributes since it seems that both the change in dkim status and domain verification have eventual consistency issues so sometimes the subsequent calls get the wrong status if we don't retry
2. A hack to force waiting for an initial state in tests because it seems even after waiting for the correct status, subsequent calls can return the wrong status. I think this won't be an issue for reall use because surely it's a temporary thing after a change and eventually the APIs start returning consistent again.
3. Cleanup of the domain identity for the email domain that's created by the verfy_domain_dkim call (in addition to the actual email identity) so that the tests cleanup everything they create.

I think it makes more sense to continue with this PR rather than me raising one. I know how to deal with crazy API inconsistencies, but you're much better placed to deal with the module in general since I don't know about DKIM signing.

So have a look over the addiitonal commits and pull them in if they make sense to you.

I thinkk the policy is to require PRs to have either a single author or one commit per author to avoid issues when squashing, so you'll probalby have to squash your commits into one and the same for mine. I've left as seperte commits for now to make it easier to follow the changes.

**Comment by stefanhorning** [comment]
Thanks for dealing with this, will have a look at your changes and probably pull them in the next couple of days or so.

**Comment by stefanhorning** [comment]
@orthanc Pulled in your changes now using this PR https://github.com/mediapeers/ansible/pull/3

**Comment by stefanhorning** [comment]
Somehow Shippable randomly cancels individual test runs, will try to rebase this branch to improve the situation as all tests that do run are green by now.

**Comment by stefanhorning** [comment]
@orthanc I manged to fix some minor issues after merging in your code. But there is still an issue in check-mode.
I know what the issue is, but not sure where to fix it best. After adding check mode behaviour to to the `get_identity_dkim_settings` method, I only return and empty dict for dkim_attributes which fails in the `_get_state_check` method (and potentially elsewhere).

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (stefanhorning) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:121]: Leaving `dkim_attributes` empty in check mode at the moment. Or do you thinks it's better to set a dummy dict here? Like `dkim_attributes = { 'DkimEnabled': true ... }`? I think not, as that would be potentially lying about the result...

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (stefanhorning) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:196]: So, I guess we have to fix it here to allow 'any' state while in `check_mode`. Simillar as in line 216.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (stefanhorning) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:109]: OR somehow catch key errors here and return s.th. instead...

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (orthanc) [inline test/integration/targets/aws_ses_identity_dkim/tasks/main.yml]: I don't think this is correct (I should have renamed the task). In general check mode should return "as if it made the change". Doing the actually attributes for this tends to be hard but at least the changed flag should represent whether a change whould have been made.

So if the task would have made a change if it were not in check mode, it should return True from changed even though it didn't actually make a change.

That way you can use check mode to see  what changes would be made.

**Comment by orthanc** [comment]
Thanks for fixing those linting issues, I thought I'd run the right checks but have missed a couple.

I need to have a proper look over your changes, but the general principle of check mode is "it should return as if it made the changes". So the return of a task run in check mode should ideally be the same as run without checkmode, but without actually making any changes.

The theoretical ideal is that you can run the whole playbook in check mode to see what would change without making any actual changnes.

So what we should be aiming for is:
* if the module would change something if it were not in check mode, then it should report changed: True
* where pragmatically possible it's nice to return the attributes as if the change was made, so yes, lying about DKIM Enabled is probably the right thing to do.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:196]: We shouldn't need to (but perhaps I'm missing something here). The state_check function here is used for two purposes:

1. Checking whether the identity is already in the right state to decide if we need to make any changes.
2. Waiting until the get_dkim_attributes call returns the right state if we do make any changes

The first should look for the actual state regardless of whether we're in check mode or not because we want to indicate whether or not we would change something.

The second needs to not wait in check mode because no change was made so it would wait for nothing.

Now that I think about it though, there's actually no need to call get_dkim_attributes again on line 218 in check_mode since we didn't actually make any change, so we could replace that wth something like what you've done on line 120ish.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:109]: Not quite sure what you mean here, can you expand a bit?

What I'm trying to do with the any case is to allow for a "don't wait" case in get_dkim_attributes when we're not waiting for a particular state without having a special case. But it may be clearer to just say wait_for=None means don't wait and get rid of the any case here.

File: test/integration/targets/aws_ses_identity_dkim/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_ses_identity_dkim/tasks/main.yml]: Ok, then I got the check mode purpose slightly wrong.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (stefanhorning) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:109]: If we don't have the key `DkimEnabled` (see `dkim_attributes` comment above) we get a key error here.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:121]: This function is being used both for getting the initial state as well as for getting the state of DKIM after a change.

I think the change her to not call get_identty_dkim in check mode is intended for the after the change call, but it's going to cause problems in the initial check about whether a change is needed.

As you've identified it's causing errors in the check, but also since the point of check mode is to check if a change would be made we really do need to call the real get_identity_dkim_attributes even if check mode is enabled.

So I'd suggest switching this back to caling get_identity_dkim attribute regardless of check_mode. The check_mode can be better handled in the main method, see comment below.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:109]: I get it. I think the answer is actually to leave this unchanged and make get_identity_dkim_attributes always call the real method even if we're in check mode (see comment on that method).

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_dkim.py:217]: This is where we should have the special handling of the attributes in check_mode rather than in get_identity_dkim attributes since we only want to skip the call after the change not the initial state check. So 215-217 should become:

    if module.check_mode:
        attributes = {
            'DkimEnabled': state == 'enabed'
        }
    else:
        attributes = get_identity_dkim_settings(module, client, identity, wait_for)

I.e.

* when not check mode, call get_identity_dkim_settings and wait for it to be in the correct state.
* when check mode, just create the expected attribute structure as if we made the change.

**Review by orthanc — CHANGES_REQUESTED** [review]
@stefanhorning Sorry it's taken me a while to get back to this, been a busy week.

I've added a few comments inline, thanks for clarifying what you mean with the NoSuchAttributeError, I get it now.

To make check_mode work properly we do need to call get_dkim_identity_attributes and get the real state so we know if a change would / wouldn't be made. So we should only fake out the call to get the new state after the change, not all usages of get_dkim_identity_attributes. That should fix the error as well as the behaviour.

More detail on this in the inline comments.

**Comment by stefanhorning** [comment]
I will close this PR as it will never be merged here, might open a fresh one in the new community.aws repo.

**Comment by stefanhorning** [comment]
New PR was opened here https://github.com/ansible-collections/community.aws/pull/71

**Pull request closed without merging**
